nomad/job_endpoint_validators.go

```diff
@@ -0,0 +1,56 @@
+package nomad
+
+import (
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/pkg/errors"
+)
+
+type jobNamespaceConstraintCheckHook struct {
+	srv *Server
+}
+
+func (jobNamespaceConstraintCheckHook) Name() string {
+	return "namespace-constraint-check"
+}
+
+func (c jobNamespaceConstraintCheckHook) Validate(job *structs.Job) (warnings []error, err error) {
+	// This was validated before and matches the WriteRequest namespace
+	ns, err := c.srv.State().NamespaceByName(nil, job.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	if ns == nil {
+		return nil, errors.Errorf("requested namespace %s does not exist", job.Namespace)
+	}
+
+	for _, tg := range job.TaskGroups {
+		for _, t := range tg.Tasks {
+			if !taskValidateDriver(t, ns) {
+				return nil, errors.Errorf(
+					"used task driver '%s' in %s[%s] is not allowed in namespace %s",
+					t.Driver, tg.Name, t.Name, ns.Name)
+			}
+		}
```

Comment:
Because a job can have multiple task groups and tasks, each with their own driver, it might be nice to collect a list disallowed task drivers and return a single error for the set of them.